### PR TITLE
cmd/roachprod: fork from cockroachdb/roachprod

### PR DESCRIPTION
**Pull request by petermattis**
Doing this in an effort to track down when stop-and-wait doesn't with
the roachtest CI acceptance tests. And it is also a reasonable long term
move so that we can starting transforming roachprod into a library for
use by roachtest.

### CODE REVIEW

**Comment by petermattis** [comment]
In addition to `DO NOT MERGE`, `DO NOT REVIEW` (yet).

**Comment by petermattis** [comment]
@benesch The extra instrumentation is showing the problem:

```
  for pid in ${pids}; do
    echo "${pid}: checking" >> %[1]s/roachprod.log
    while kill -0 ${pid}; do
      kill -0 ${pid} >> %[1]s/roachprod.log 2>&1
      echo "${pid}: still alive [$?]" >> %[1]s/roachprod.log
      ps axeww -o pid -o command >> %[1]s/roachprod.log
      sleep 1
    done
    echo "${pid}: dead" >> %[1]s/roachprod.log
  done
```

The resulting `roachprod.log` shows:

```
292: checking
292: still alive [0]
...
  292 [cockroach] <defunct>
...
292: still alive [0]
...
  292 [cockroach] <defunct>
...
```

And that loop just continues forever. So we definitely killed the `cockroach` process, but the process is never being reaped. Something interesting I noticed is that PID 1 is:

`1 ./bin/roachtest run (acceptance|kv/splits) --local --cockroach cockroach --roachprod bin/roachprod --workload bin/workload --artifacts artifacts --teamcity`

Could it be that `roachtest` is becoming the parent of the `cockroach` process and not reaping it?

**Comment by petermattis** [comment]
This is ready for a look.

**Comment by petermattis** [comment]
Ping. This is going to be irritating to keep up to date.

File: pkg/cmd/roachprod/main.go [hunk 1]
```diff
@@ -118,14 +118,19 @@ func newCluster(name string, reserveLoadGen bool) (*install.SyncedCluster, error
 
 	c, ok := install.Clusters[name]
 	if !ok {
-		return nil, fmt.Errorf(`unknown cluster: %s
+		// NB: We don't use fmt.Errorf due to a linter error about the error
+		// message containing capitals and punctuation. We don't use
```
Comment (andreimatei) [inline pkg/cmd/roachprod/main.go]: I hate that linter warning about the capitalization. And does it really complain about punctuation? I feel like I put multiple sentences with periods between them in errors all the time.
Let's rip out the linter, now that Tamir is gone.

**Comment by andreimatei** [comment]
LGTM

**Review by andreimatei — APPROVED** [review]

File: pkg/cmd/roachprod/main.go (hunk 1 above)
Comment (petermattis) [reply pkg/cmd/roachprod/main.go]: Feel free to do that, though I thought this error was baked into `golint`. Perhaps I'm mistaken about that.

**Comment by petermattis** [comment]
@benesch Care to give this your rubber stamp?

**Review by benesch — APPROVED** [review]
:lgtm: :bow: :raised_hands:

> Reviewed 35 of 35 files at r12, 3 of 3 files at r13, 1 of 1 files at r14, 1 of 1 files at r15, 1 of 1 files at r16, 1 of 1 files at r17, 10 of 10 files at r18, 30 of 30 files at r19, 17 of 17 files at r21, 1 of 1 files at r22, 15 of 15 files at r23.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32267#-:-LRJEN840PujQUUihgNd:bv3qj8o)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/cmd/roachprod/main.go, line 122 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/32267#-LRCkPI-I0NB7mp_nVye-r11-122:-LRJBsakD2qp_8MVBajn:b2saek0) ([raw file](https://github.com/cockroachdb/cockroach/blob/5eccf75a5a9b20d9a66004316c7d9a3b9131fa9d/pkg/cmd/roachprod/main.go#L122)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Feel free to do that, though I thought this error was baked into `golint`. Perhaps I'm mistaken about that.
</blockquote></details>

Errors with full sentences don't compose well. I agree the `function one: function two: function three: function four: thing happened` errors aren't great, but I think `function one: function two: this error happened. Are you sure that you didn't mean --insecure?: function four: unable to open socket` are even worse.

tl;dr I don't think we should just rip out that linter. I think there's a broader conversation about errors to be had. E.g. here I think we should return a type that's interpreted by a top level CLI error interpreter that tacks on more context.
___
*[pkg/cmd/roachprod/main.go, line 228 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/32267#-LRJDusp4_Yc0Hjv_nas:-LRJDusqD67D_KIVaMZK:bj7fsau) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fe550419463a2122ea0533b325e456814278190/pkg/cmd/roachprod/main.go#L228)):*
> ```Go
> 		// account is "peter", `roachprod create perf` should be specified as
> 		// `roachprod create peter-perf`.
> 		_ = 0
> ```

Heh, my turn to complain: this is my least favorite linter.
___
*[pkg/cmd/roachprod/install/cockroach.go, line 386 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/32267#-LRJDDkMAF1zxrVUVLFZ:-LRJDDkMAF1zxrVUVLF_:b-czbkh) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b71bcda2a0d40ba20894ac07b7b3d3bfa4c8593/pkg/cmd/roachprod/install/cockroach.go#L386)):*
> ```Go
> 
> 	if bootstrapped {
> 		license := envutil.EnvOrDefaultString("COCKROACH_DEV_LICENSE", "")
> ```

huh, I didn't think the envutil linter was active for pkg/cmd. Weird.

**Comment by petermattis** [comment]
Oops, forgot to push the updated vendor repo.

**Review by petermattis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32267)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/cmd/roachprod/main.go, line 122 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/32267#-LRCkPI-I0NB7mp_nVye-r11-122:-LRJtiMvDYBus9DIfCW4:bsdtae5) ([raw file](https://github.com/cockroachdb/cockroach/blob/5eccf75a5a9b20d9a66004316c7d9a3b9131fa9d/pkg/cmd/roachprod/main.go#L122)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Errors with full sentences don't compose well. I agree the `function one: function two: function three: function four: thing happened` errors aren't great, but I think `function one: function two: this error happened. Are you sure that you didn't mean --insecure?: function four: unable to open socket` are even worse.

tl;dr I don't think we should just rip out that linter. I think there's a broader conversation about errors to be had. E.g. here I think we should return a type that's interpreted by a top level CLI error interpreter that tacks on more context.
</blockquote></details>

@benesch You're right that there is a possible refactoring here that would make this cleaner. I'm leaving this as-is for now. You and @andreimatei can fight it out as to whether removing the linter or refactoring is the better option.
___
*[pkg/cmd/roachprod/main.go, line 228 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/32267#-LRJDusp4_Yc0Hjv_nas:-LRJtuKh2WWdzjpREq4I:b2byjin) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fe550419463a2122ea0533b325e456814278190/pkg/cmd/roachprod/main.go#L228)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Heh, my turn to complain: this is my least favorite linter.
</blockquote></details>

Yeah, this is ridiculous. I like being able to comment on the else branch. That was intentional, Mr. Linter.
___
*[pkg/cmd/roachprod/install/cockroach.go, line 386 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/32267#-LRJDDkMAF1zxrVUVLFZ:-LRJuFVk1zzYVGDlOOvo:bkzyknf) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b71bcda2a0d40ba20894ac07b7b3d3bfa4c8593/pkg/cmd/roachprod/install/cockroach.go#L386)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

huh, I didn't think the envutil linter was active for pkg/cmd. Weird.
</blockquote></details>

You're right, but something seems to be broken as I definitely got a warning. I'm going to get this merged, but investigate what is going wrong there in parallel.

**Review by petermattis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32267)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/cmd/roachprod/install/cockroach.go, line 386 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/32267#-LRJDDkMAF1zxrVUVLFZ:-LRJwMFfA3eBe9qqfh6M:b-oxeqta) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b71bcda2a0d40ba20894ac07b7b3d3bfa4c8593/pkg/cmd/roachprod/install/cockroach.go#L386)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

You're right, but something seems to be broken as I definitely got a warning. I'm going to get this merged, but investigate what is going wrong there in parallel.
</blockquote></details>

Found it: the linter doesn't exclude `pkg/cmd` for the pattern `\bos\.(Getenv|LookupEnv)\("COCKROACH`. It does exclude `pkg/cmd` for the pattern `\bos\.(Getenv|LookupEnv)\(`.

**Comment by petermattis** [comment]
bors r=andreimatei,benesch

**Review by benesch — COMMENTED** [review]
:plause: :sparkly_cockroach: on this heroic effort

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32267#-:-LRKejef9e2mKXVkB_Mh:b-tv1waa)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/cmd/roachprod/install/cockroach.go, line 386 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/32267#-LRJDDkMAF1zxrVUVLFZ:-LRKeiUO7-3esLOk-dQK:b-fj8rsq) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b71bcda2a0d40ba20894ac07b7b3d3bfa4c8593/pkg/cmd/roachprod/install/cockroach.go#L386)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Found it: the linter doesn't exclude `pkg/cmd` for the pattern `\bos\.(Getenv|LookupEnv)\("COCKROACH`. It does exclude `pkg/cmd` for the pattern `\bos\.(Getenv|LookupEnv)\(`.
</blockquote></details>

Ha! Nice spot.
